File: MATLAB_edition/frame3d_eg.m
%立体刚架元算例。

s=xlsread('frame3d_eg_input.xlsx');
s(all(isnan(s),2),:)=[];
eles=s(1,1);
nodes=s(1,2);
e=s(2,1:eles);
g=s(3,1:eles);
a=s(4,1:eles);
iy=s(5,1:eles);
iz=s(6,1:eles);
j=s(7,1:eles);
node_number=s(8:7+eles,1:2)';   %每个单元对应节点的编号。
node_coordinate=s(8+eles:7+eles+nodes,1:3)';
node_coordinate=truss3d_ele_coordinate(node_coordinate,node_number);   %每个单元对应节点的坐标。
unknown_u_index=s(8+eles+nodes,:);
unknown_u_index(:,all(isnan(unknown_u_index),1))=[];
unknown_u_index=unknown_u_index';
f_known=s(9+eles+nodes,1:size(unknown_u_index,1))';
u_known=s(10+eles+nodes,1:6*nodes-size(unknown_u_index,1))';

ele_info=zeros(10,eles);    %单元信息矩阵，每列依次是e\a\iy\iz\j\l\cXx\cYx\cZx\g。
ele_info(1,:)=e;
ele_info(2,:)=a;
ele_info(3,:)=iy;
ele_info(4,:)=iz;
ele_info(5,:)=j;
ele_info(10,:)=g;
ele_info(6:9,:)=frame3d_ele_info(node_coordinate);  %补全单元信息矩阵的信息。

r=zeros(12,12,eles);   %r为单元转移矩阵的集合，每单元十二行12列。
for i=1:1:eles
    r(:,:,i)=frame3d_r_mat(ele_info(7:9,i));
end     %完成单元转移矩阵的集合。

k_ele=zeros(12,12,eles);  %k_ele是单元在局部坐标系下的刚度矩阵，每单元十二行12列。
for i=1:1:eles
    k_ele(:,:,i)=frame3d_k_ele_mat(ele_info(:,i));
end     %完成局部坐标系下的单元刚度矩阵集合。

kk=zeros(12,12,eles); %kk是整体坐标下单元刚度矩阵的集合，每单元12行12列。
for i=1:1:eles
    kk(:,:,i)=frame3d_ele_stiff_mat(r(:,:,i),k_ele(:,:,i));
end     %完成整体坐标系下单元刚度矩阵集合。


k=zeros(nodes*6,nodes*6);   %k是总体刚度矩阵。
for i=1:1:eles
    k=frame3d_assembly_mat(k,kk(:,:,i),node_number(:,i));
end     %w完成总体刚度矩阵。

unknown_var=frame3d_solve_unknown(k,unknown_u_index,f_known,u_known,nodes);     %利用分块矩阵计算未知的变量，力和位移。
ua = unknown_var{1,1};
ua=double(ua);
fc = unknown_var{1,2};
fc=double(fc);
U = unknown_var{1,3};
F = unknown_var{1,4};
kcc = unknown_var{1,5};
kca = unknown_var{1,6};
kac = unknown_var{1,7};
kaa = unknown_var{1,8};

u=zeros(12,eles);   %u为每单元两节点的位移阵列。
for i=1:1:eles
   u(:,i)=[U(node_number(1,i)*6-5:node_number(1,i)*6);U(node_number(2,i)*6-5:node_number(2,i)*6)];
end

f_nodal=zeros(12,eles);    %f为每单元两节点的受力阵列，按以下对每个单元操作的步骤后，还应减去等效节点载荷。
for i=1:1:eles
    f_nodal(:,i)=k_ele(:,:,i)*r(:,:,i)*u(:,i);
end


%以下写文件。
f=fopen('frame3d_eg_output.xls','w');

fprintf(f, 'element stiffness matrices:\r\n\r\n');
for i=1:eles
    for j=1:12
        for m=1:12
            fprintf(f,'%8.4f\t',kk(j,m,i));
        end
        fprintf(f, '\r\n');
    end
    fprintf(f,'\r\n');
end

fprintf(f,'\r\nglobal stiffness matrix:\r\n\r\n');
for j=1:6*nodes
    for m=1:6*nodes
        fprintf(f,'%8.4f\t',k(j,m));
    end
    fprintf(f, '\r\n');
end

fprintf(f,'\r\n\r\nK_cc= \r\n\r\n');
for i=1:size(kcc,1)
    for j=1:size(kcc,2)
        fprintf(f,'%8.4f\t',kcc(i,j));
    end
    fprintf(f,'\r\n');
end
fprintf(f,'\r\n\r\nK_ca= \r\n\r\n');
for i=1:size(kca,1)
    for j=1:size(kca,2)
        fprintf(f,'%8.4f\t',kca(i,j));
    end
    fprintf(f,'\r\n');
end
fprintf(f,'\r\n\r\nK_ac= \r\n\r\n');
for i=1:size(kac,1)
    for j=1:size(kac,2)
        fprintf(f,'%8.4f\t',kac(i,j));
    end
    fprintf(f,'\r\n');
end
fprintf(f,'\r\n\r\nK_aa= \r\n\r\n');
for i=1:size(kaa,1)
    for j=1:size(kaa,2)
        fprintf(f,'%8.4f\t',kaa(i,j));
    end
    fprintf(f,'\r\n');
end

fprintf(f,'\r\n\r\nU_a= \r\n\r\n');
for i=1:size(ua,1)
    fprintf(f,'%8.4f\t',ua(i));
end
fprintf(f,'\r\n\r\nF_c= \r\n\r\n');
for i=1:size(fc,1)
    fprintf(f,'%8.4f\t',fc(i));
end
fprintf(f,'\r\n\r\nU= \r\n\r\n');
for i=1:size(U,1)
    fprintf(f,'%8.4f\t',U(i));
end
fprintf(f,'\r\n\r\nF= \r\n\r\n');
for i=1:size(F,1)
    fprintf(f,'%8.4f\t',F(i));
end

fprintf(f,'\r\n\r\nf_nodal= \r\n\r\n');
for i=1:size(f_nodal,1)
    for j=1:size(f_nodal,2)
        fprintf(f,'%8.4f\t\t',f_nodal(i,j));
    end
    fprintf(f,'\r\n');
end

fclose(f);




































% e=210e9;
% g=84e9;
% a=2e-2;
% iy=1e-4;
% iz=2e-4;
% j=5e-5;
% eles=8;
% nodes=8;
% node_number=[1,2,3,4,5,6,7,8;
%     5,6,7,8,6,7,8,5];  %每个单元对应节点的编号。
% node_coordinate=[0,0,0,0,5,0;
%                  0,0,4,0,5,4;
%                  4,0,4,4,5,4;
%                  4,0,0,4,5,0;
%                  0,5,0,0,5,4;
%                  0,5,4,4,5,4;
%                  4,5,4,4,5,0;
%                  4,5,0,0,5,0]';  %每个单元对应节点的坐标。
% ele_info=zeros(10,eles);    %单元信息矩阵，每列依次是e\g\a\iy\iz\j\l\cXx\cYx\cZx。
% ele_info(1,:)=e*ones(1,eles);
% ele_info(2,:)=a*ones(1,eles);
% ele_info(3,:)=iy*ones(1,eles);
% ele_info(4,:)=iz*ones(1,eles);
% ele_info(5,:)=j*ones(1,eles);
% ele_info(10,:)=g*ones(1,eles);
% ele_info(6:9,:)=frame3d_ele_info(node_coordinate);  %补全单元信息矩阵的信息。
% r=zeros(12*eles,12);   %r为单元转移矩阵的集合，十二eles行12列。
% for i=1:1:eles
%     r(12*i-11:12*i,:)=frame3d_r_mat(ele_info(7:9,i));
% end     %完成单元转移矩阵的集合。
% k_ele=zeros(12*eles,12);  %k_ele是单元在局部坐标系下的刚度矩阵，十二eles行12列。
% for i=1:1:eles
%     k_ele(12*i-11:12*i,:)=frame3d_k_ele_mat(ele_info(:,i));
% end     %完成局部坐标系下的单元刚度矩阵集合。
% kk=zeros(12*eles,12); %kk是整体坐标下单元刚度矩阵的集合，12eles行12列。
% for i=1:1:eles
%     kk(12*i-11:12*i,:)=frame3d_ele_stiff_mat(r(12*i-11:12*i,:),k_ele(12*i-11:12*i,:));
% end     %完成整体坐标系下单元刚度矩阵集合。
% k=zeros(nodes*6,nodes*6);   %k是总体刚度矩阵。
% for i=1:1:eles
%     k=frame3d_assembly_mat(k,kk(12*i-11:12*i,:),node_number(:,i));
% end     %w完成总体刚度矩阵。
%  m=k(25:48,25:48);
%  f=[zeros(12,1);-15e3;zeros(11,1)];
%  u=m\f;
%  U=[zeros(24,1);u]   %U为节点总位移列阵。
%  F=k*U   %F为节点总受力列阵。
%  u=zeros(12,eles);   %u为每单元两节点的位移阵列。
%  for i=1:1:eles
%     u(:,i)=[U(node_number(1,i)*6-5:node_number(1,i)*6);U(node_number(2,i)*6-5:node_number(2,i)*6)];
%  end
% f=zeros(12,eles);    %f为每单元两节点的受力阵列，按以下对每个单元操作的步骤后，还应减去等效节点载荷。
% for i=1:1:eles
%     f(:,i)=k_ele(12*i-11:12*i,:)*r(12*i-11:12*i,:)*u(:,i);
% end
% f

File: Python_edition/frame3d_eg/frame3d_eg.py
import numpy as np
from math import *
import pandas as pd
import matplotlib.pyplot as plt
from xlwt import *

from fem.frame3d_ele_coordinate import *
from fem.frame3d_ele_info import *
from fem.frame3d_r_mat import *
from fem.frame3d_k_ele_mat import *
from fem.frame3d_ele_stiff_mat import *
from fem.frame3d_assembly_mat import *
from fem.frame3d_solve_unknown import *


def main():
    # -----------------数据的读入与有效提取。--------------------------------------------------------------------------#
    a = pd.read_excel('frame3d_eg_input.xlsx', sheet_name='Sheet1')
    b = a
    for i in a.index:
        if type(a.iloc[i, 0]) == str:
            b = b.drop([i])
    s = b.values  # s为处理后得到的源数据。之后再进行提取。

    eles = s[0, 0]  # 单元个数。
    nodes = int(s[0, 1])  # 节点个数。
    dof = int(s[0, 2])  # 单个节点的自由度。
    n_for_e = int(s[0, 3])  # 每个单元的节点个数。
    size_ele_mat = dof * n_for_e  # 单元刚度矩阵的规模。
    e = s[1, 0:eles]
    g = s[2, 0:eles]
    a = s[3, 0:eles]
    iy = s[4, 0:eles]
    iz = s[5, 0:eles]
    j = s[6, 0:eles]  # 以上各元素，分别对应弹性模量、剪切模量、面积、两轴的惯性矩、极惯性矩。

    node_number = np.array(s[7:7 + eles, 0:2].transpose(), dtype=int)  # 每个单元对应节点的编号。
    node_coordinate = s[7 + eles:7 + eles + nodes, 0: 3].transpose()
    # node_coordinate为各单元节点对应的坐标。
    node_coordinate = frame3d_ele_coordinate(node_coordinate, node_number)

    # unknown_u_index为未知位移节点的编号。
    unknown_u_index = s[7 + eles + nodes, :]
    unknown_u_index = np.array(unknown_u_index[~pd.isnull(unknown_u_index)], dtype=int)
    unknown_u_index.shape = (unknown_u_index.shape[0], 1)

    # f_known为已知的节点力。
    f_known = s[8 + eles + nodes, 0:unknown_u_index.shape[0]]
    f_known = f_known[~pd.isnull(f_known)]
    f_known.shape = (f_known.shape[0], 1)

    # u_known为已知的节点位移。
    u_known = s[9 + eles + nodes, 0:dof * nodes - unknown_u_index.shape[0]]
    u_known = u_known[~pd.isnull(u_known)]
    u_known.shape = (u_known.shape[0], 1)

    # # 以下完成单元信息矩阵，每列依次是e\a\iy\iz\j\l\cXx\cYx\cZx\g。
    ele_info = np.zeros((10, eles), dtype=float)
    ele_info[0, :] = e
    ele_info[1, :] = a
    ele_info[2, :] = iy
    ele_info[3, :] = iz
    ele_info[4, :] = j
    ele_info[9, :] = g
    ele_info[5: 9, :] = frame3d_ele_info(node_coordinate)

    # -----------------------单元刚度矩阵的建立。-----------------------------------------------------------------------#

    # r为单元转移矩阵的集合，每单元六行六列。
    r = np.zeros((eles, size_ele_mat, size_ele_mat))
    for i in range(eles):
        r[i, :, :] = frame3d_r_mat(ele_info[6: 9, i])

    # k_ele是单元在局部坐标系下的刚度矩阵，每单元六行六列。
    k_ele = np.zeros((eles, size_ele_mat, size_ele_mat))
    for i in range(eles):
        k_ele[i, :, :] = frame3d_k_ele_mat(ele_info[:, i])

    # kk是整体坐标下单元刚度矩阵的集合，每单元六行六列。
    kk = np.zeros((eles, size_ele_mat, size_ele_mat))
    for i in range(eles):
        kk[i, :, :] = frame3d_ele_stiff_mat(r[i, :, :], k_ele[i, :, :])

    # ----------------------总体方程的求解。----------------------------------------------------------------------------#
    # k是总体刚度矩阵。
    k = np.zeros((nodes * dof, nodes * dof))
    for i in range(eles):
        k = frame3d_assembly_mat(k, kk[i, :, :], node_number[:, i], size_ele_mat)

    # 利用分块矩阵计算未知的变量，力和位移。
    unknown_var = frame3d_solve_unknown(k, unknown_u_index, f_known, u_known, nodes, dof)
    ua = np.array(unknown_var[0], dtype=float)
    fc = np.array(unknown_var[1], dtype=float)
    U = np.array(unknown_var[2], dtype=float)
    F = np.array(unknown_var[3], dtype=float)
    kcc = np.array(unknown_var[4], dtype=float)
    kca = np.array(unknown_var[5], dtype=float)
    kac = np.array(unknown_var[6], dtype=float)
    kaa = np.array(unknown_var[7], dtype=float)

    # ---------------------------后处理。-------------------------------------------------------------------------------#

    # u为每单元两节点的位移阵列。
    u = np.zeros((size_ele_mat, eles), dtype=float)
    for i in range(eles):
        u[0:6, i] = np.array(U[node_number[0, i] * dof - dof: node_number[0, i] * dof, 0], dtype=float)
        u[6:12, i] = np.array(U[node_number[1, i] * dof - dof: node_number[1, i] * dof, 0], dtype=float)

    # f为每单元两节点的受力阵列，按以下对每个单元操作的步骤后，还应减去等效节点载荷。
    f_nodal = np.zeros((size_ele_mat, eles), dtype=float)
    for i in range(eles):
        f_nodal[:, i] = np.dot(np.dot(k_ele[i, :, :], r[i, :, :]), u[:, i])

    # =============文件写入。============================================================================================

    wb = Workbook(encoding='utf-8')
    sh = wb.add_sheet('frame_3d')

    sh.write(0, 0, 'element stiffness matrices:')
    for i in range(eles):
        for j in range(size_ele_mat):
            for m in range(size_ele_mat):
                sh.write(i * (size_ele_mat + 1) + j + 1, m, kk[i, j, m])

    sh.write((size_ele_mat + 1) * eles + 1, 0, 'global stiffness matrix:')
    for j in range(dof * nodes):
        for m in range(dof * nodes):
            sh.write((size_ele_mat + 1) * eles + 2 + j, m, k[j, m])

    sh.write((size_ele_mat + 1) * eles + dof * nodes + 3, 0, 'K_cc=')
    for i in range(kcc.shape[0]):
        for j in range(kcc.shape[1]):
            sh.write((size_ele_mat + 1) * eles + dof * nodes + 4 + i, j, kcc[i, j])

    sh.write((size_ele_mat + 1) * eles + dof * nodes + 5 + kcc.shape[0], 0, 'K_ac=')
    for i in range(kac.shape[0]):
        for j in range(kac.shape[1]):
            sh.write((size_ele_mat + 1) * eles + dof * nodes + 6 + kcc.shape[0] + i, j, kac[i, j])

    sh.write((size_ele_mat + 1) * eles + 2 * dof * nodes + 7, 0, 'K_ca=')
    for i in range(kca.shape[0]):
        for j in range(kca.shape[1]):
            sh.write((size_ele_mat + 1) * eles + 2 * dof * nodes + 8 + i, j, kca[i, j])

    sh.write((size_ele_mat + 1) * eles + 2 * dof * nodes + 9 + kca.shape[0], 0, 'K_aa=')
    for i in range(kaa.shape[0]):
        for j in range(kaa.shape[1]):
            sh.write((size_ele_mat + 1) * eles + 2 * dof * nodes + 10 + kca.shape[0] + i, j, kaa[i, j])

    sh.write((size_ele_mat + 1) * eles + 3 * dof * nodes + 11, 0, 'U_a=')
    for i in range(ua.shape[0]):
        sh.write((size_ele_mat + 1) * eles + 3 * dof * nodes + 12, i, ua[i, 0])

    sh.write((size_ele_mat + 1) * eles + 3 * dof * nodes + 14, 0, 'F_c=')
    for i in range(fc.shape[0]):
        sh.write((size_ele_mat + 1) * eles + 3 * dof * nodes + 15, i, fc[i, 0])

    sh.write((size_ele_mat + 1) * eles + 3 * dof * nodes + 17, 0, 'U=')
    for i in range(U.shape[0]):
        sh.write((size_ele_mat + 1) * eles + 3 * dof * nodes + 18, i, U[i, 0])

    sh.write((size_ele_mat + 1) * eles + 3 * dof * nodes + 20, 0, 'F=')
    for i in range(U.shape[0]):
        sh.write((size_ele_mat + 1) * eles + 3 * dof * nodes + 21, i, F[i, 0])

    sh.write((size_ele_mat + 1) * eles + 3 * dof * nodes + 23, 0, 'f_nodal=')
    for i in range(f_nodal.shape[0]):
        for j in range(f_nodal.shape[1]):
            sh.write((size_ele_mat + 1) * eles + 3 * dof * nodes + 24 + i, 2 * j, f_nodal[i, j])

    wb.save('frame3d_eg_output_py.xls')


# =======================================================================================================================
if __name__ == '__main__':
    main()
